### PR TITLE
Add lsbfdump example using Scala and Rust

### PR DESCRIPTION
**Pull request by shanedell**

### CODE REVIEW

File: lsbfdump-jni-rs/build.sbt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+lazy val root = project
+  .in(file("."))
+  .settings(commonSettings)
+  .settings(
+    name := "lsbfdump-jni-rs",
+    Compile / mainClass := Some("org.apache.daffodil.lsbfDump.LSBFDump"),
+    libraryDependencies ++= List(
+      "junit" % "junit" % "4.13.2" % Test,
+      // needed below package to successfully run junit tests
+      "com.novocode" % "junit-interface" % "0.11" % Test exclude("junit", "junit-dep")
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/build.sbt]: This is com.github.sbt not com.novocode and version 0.13.3 is available.

File: lsbfdump-jni-rs/native/Cargo.lock
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/Cargo.lock:1]: lock file should not be checked in. Add to the .gitignore.

File: lsbfdump/build.sbt [hunk 2]
```diff
@@ -16,5 +16,7 @@ nativeImageOptions ++= Seq(
 
 // Library dependencies
 libraryDependencies ++= Seq(
+  // needed below package to successfully run junit tests
+  "com.novocode" % "junit-interface" % "0.11" % Test exclude("junit", "junit-dep"),
```
Comment (mbeckerle) [inline lsbfdump/build.sbt]: This is com.github.sbt not com.novocode and version 0.13.3 is available.

File: lsbfdump-jni-rs/native/src/helper_functions.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    }
+  }
+
+  Ok(lines)
+}
+
+pub fn open_and_seek_input_stream(filename: &str, offset: u64) -> io::Result<Box<dyn CustomReadSeek>> {
+  if filename == "-" {
+    let mut stdin_with_seek = StdinWithSeek;
+    stdin_with_seek.seek(SeekFrom::Start(offset))?;
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/src/helper_functions.rs]: ok, I learned a bunch of Rust trait stuff from this, but wouldn't it be simpler right here to just advance the stdin by the offset number of bytes here, then you don't need any of the CustomReadSeek trait stuff at all. No more seeks are needed except this one, so I think you would not need a `<dyn ...>` either.

File: .gitignore [hunk 4]
```diff
@@ -41,3 +41,8 @@ resource_managed
 
 # Project local cache of managed dependencies
 lib_managed
+
+# vscode/metals generated files
+.vscode
+.metals
+.bloop
```
Comment (mbeckerle) [inline .gitignore:48]: Add Cargo.lock here I think.

File: lsbfdump-jni-rs/native/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[package]
+name = "lsbfdump"
+version = "0.1.0"
+authors = ["Shane Dell <shanedell100@gmail.com>"]
+edition = "2021"
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/Cargo.toml:5]: 2021 ??

File: lsbfdump-jni-rs/native/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+          eprintln!("lsbfdump: An I/O error occurred - {}", err);
+          // std::process::exit(1);
+          return false as jboolean
+        }
+      }
+    }
+    Err(err) => {
+      eprintln!("lsbfdump: An I/O error occurred - {}", err);
+      // std::process::exit(1);
+      return false as jboolean
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/src/lib.rs:94]: Ok, so all the arguments are simple things here, and the return is a simple jboolean. 

If the args or returns are data structures, then we're into marshalling to/from some string or byte representation yes?

**Review by mbeckerle — APPROVED** [review]
+1
Very interesting. I made a few comments, asked a few questions.

File: lsbfdump-jni-rs/native/src/helper_functions.rs (hunk 3 above)
Comment (shanedell) [reply lsbfdump-jni-rs/native/src/helper_functions.rs]: I made these updates, thanks for that catch!

File: lsbfdump-jni-rs/native/Cargo.toml (hunk 5 above)
Comment (shanedell) [reply lsbfdump-jni-rs/native/Cargo.toml:5]: Yeah so I just made this using `cargo new <project-name>`. That is what it auto sets  `edition` to, if you try changing it, it will most likely cause things to break. The only valid values I could find for `edition` was 2015, 2018 and 2021.

File: lsbfdump-jni-rs/native/src/lib.rs (hunk 6 above)
Comment (shanedell) [reply lsbfdump-jni-rs/native/src/lib.rs:94]: Possibly, I believe there are things like `jni::objects::JObject` as well. So something like that may also be able to be done using different `jni::objects`.

File: lsbfdump-jni-rs/native/Cargo.toml (hunk 5 above)
Comment (stevedlawrence) [reply lsbfdump-jni-rs/native/Cargo.toml:5]: Note that `edition` isn't supposed to be the year this application was made, which I think Mike was implying. It's basically the version of the rust language to use. It's like saying to use C89, C99, C11, etc. for the C language. Same core language, but different capabilities/keywords/etc.

https://doc.rust-lang.org/edition-guide/editions/

File: lsbfdump-jni-rs/native/src/helper_functions.rs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        count += 4;
+      }
+      Err(_) => break,
+    }
+  }
+
+  Ok(lines)
+}
+
+pub fn open_and_seek_input_stream(filename: &str, offset: u64) -> io::Result<Box<dyn Read>> {
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/src/helper_functions.rs:44]: I'm curious about dyn here. Why is that needed? I would think Read would be a common trait to both the StdIn and a io::BufReader.

File: lsbfdump-jni-rs/native/src/stdin_seek.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+impl Read for StdinWithSeek {
+  fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+    io::stdin().read(buf)
+  }
+}
+
+impl Seek for StdinWithSeek {
+  fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+    let mut cursor = Cursor::new(Vec::new());
+    cursor.seek(pos)
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/src/stdin_seek.rs]: I don't understand this Cursor usage. 

Or rather, if I understand this right, this will create a buffer from the input, and then seek to a position in it, filling it with data from the stdin. 

That's undesirable because if the data being streamed in is huge, and the seek position is far in, you'll allocate all that data. 

Just calling read on the stdin in a loop will provide better streaming behavior. You can call that loop in the open-and-seek-input-stream method. Then this whole file is not needed. 

Do I have that right? 

re: rust - I can't figure out from the online doc for Cursor how the Cursor is being associated to the input stream here. Is Cursor a mixin of some sort? I.e., somehow self is bound to the input stream?

File: lsbfdump-jni-rs/native/src/helper_functions.rs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  if filename == "-" {
+    let mut stdin_with_seek = StdinWithSeek;
```
Comment (mbeckerle) [inline lsbfdump-jni-rs/native/src/helper_functions.rs]: You could just call read in a loop N times to carry out the seeking right here. I don't see the need for StdinWithSeek at all. See comment later about buffering vs. streaming behavior also. 

If you do that, then I don't see a reason for a dyn anything here. It would just be a `Box<Read>` wouldn't it?

File: lsbfdump-jni-rs/native/src/helper_functions.rs (hunk 7 above)
Comment (shanedell) [reply lsbfdump-jni-rs/native/src/helper_functions.rs:44]: So the `dyn` is needed even after updating the code to loop over read. Whenever I would try to remove it, I would get an error `add 'dyn' keyword before this trait: 'dyn'`. I found online it is required, below is a quote from an answer found at https://users.rust-lang.org/t/why-does-rustc-not-automatically-infer-dyn-for-trait-objects/98535:

```
The dyn was added, and later made mandatory in the 2021 edition 1 because otherwise
it is very hard for a programmer reading the code or its documentation to know what is a
concrete type and what is a trait denoting a trait object. Being explicit about things like this
is an important part of Rust's style.
```

File: lsbfdump-jni-rs/native/src/helper_functions.rs (hunk 9 above)
Comment (shanedell) [reply lsbfdump-jni-rs/native/src/helper_functions.rs]: So the `dyn` is needed even after updating the code to loop over read. Whenever I would try to remove it, I would get an error `add 'dyn' keyword before this trait: 'dyn'`. I found online it is required, below is a quote from an answer found at https://users.rust-lang.org/t/why-does-rustc-not-automatically-infer-dyn-for-trait-objects/98535:

```
The dyn was added, and later made mandatory in the 2021 edition 1 because otherwise
it is very hard for a programmer reading the code or its documentation to know what is a
concrete type and what is a trait denoting a trait object. Being explicit about things like this
is an important part of Rust's style.
```

File: lsbfdump-jni-rs/native/src/stdin_seek.rs (hunk 8 above)
Comment (shanedell) [reply lsbfdump-jni-rs/native/src/stdin_seek.rs]: So according to https://doc.rust-lang.org/std/io/struct.Cursor.html,

> A Cursor wraps an in-memory buffer and provides it with a [Seek](https://doc.rust-lang.org/std/io/trait.Seek.html) implementation.
> Cursors are used with in-memory buffers, anything implementing [AsRef](https://doc.rust-lang.org/std/convert/trait.AsRef.html)<[u8]>, to allow them to implement [Read](https://doc.rust-lang.org/std/io/trait.Read.html) and/or [Write](https://doc.rust-lang.org/std/io/trait.Write.html), allowing these buffers to be used anywhere you might use a reader or writer that does actual I/O."

But I did implement what you mentioned about looping over read on the stdin.

File: lsbfdump-jni-rs/native/src/helper_functions.rs (hunk 9 above)
Comment (mbeckerle) [reply lsbfdump-jni-rs/native/src/helper_functions.rs]: So dyn is required if you are handling an object as an instance of a Trait or base class vs. as the actual class object.

Ok. Seems to fly in the face of OOP principles, but Ok.

File: lsbfdump-jni-rs/native/src/stdin_seek.rs (hunk 8 above)
Comment (mbeckerle) [reply lsbfdump-jni-rs/native/src/stdin_seek.rs]: Ok. I am learning a lot here. 

So Read and Write are operations on abstract buffers. Cursor makes a buffer have seek. 

And, oh, btw: input streams and output streams are buffers because they support Read and Write. 

Not the same pattern of OO factoring as other languages, but works equally well. Maybe better. In Rust there is no need to construct ByteBufferInputStream that simulates an InputStream using a memory buffer. They've flipped it. Buffer is the primitive here, and I/O Streams are Buffers. 

Works for me.
